src/Processors/QueryPlan/Optimizations/optimizeLazyMaterialization.cpp

```diff
@@ -264,6 +264,11 @@ void optimizeLazyMaterialization(QueryPlan::Node & root, Stack & stack, QueryPla
     if (lazily_read_info->lazily_read_columns.empty())
         return;
 
+    /// avoid applying this optimization on impractical queries for sake of implementation simplicity
+    /// i.e. when no column used in query until projection, for example, select * from t order by rand() limit 10
+    if (reading_step->getAllColumnNames().size() == lazily_read_info->lazily_read_columns.size())
+        return;
```

Comment:
I'm not sure this solution is optimal (if I understand it correctly) - if we have multiple columns and none of them is filtering  it still may be preferable to lazy all of them except one... again - if I understand this logic correctly...